src/session-pool.js

```diff
@@ -435,252 +366,370 @@ SessionPool.prototype.requestStream = function(config) {
 
-    var transaction = self.createTransaction_(session);
+  var idle = this.getIdleSessions_();
+  var count = idle.length;
 
-    session.isWriteSession_ = true;
-    session.transaction_ = transaction;
+  while (count-- > maxIdle && count >= min) {
+    evicted.push(idle.pop());
+  }
 
-    return transaction.begin().then(function() {
-      return session;
-    });
+  var requests = evicted.map(function(session) {
+    return self.destroySession_(session);
   });
+
+  return Promise.all(requests);
 };
 
 /**
- * Attempt to retrieve the next available session. If a write request is made,
- * but no write sessions are available, it will get a read session and begin a
- * new transaction.
+ * Retrieves a list of all the idle sessions.
  *
  * @private
  *
- * @param {object=} options - Session type options.
- * @param {boolean} options.write - Whether or not the requested session is a
- *     write session.
- * @param {function} callback - The callback function.
+ * @returns {Session[]}
  */
-SessionPool.prototype.getNextAvailableSession_ = function(options, callback) {
-  var self = this;
-
-  if (is.fn(options)) {
-    callback = options;
-    options = {};
-  }
+SessionPool.prototype.getIdleSessions_ = function() {
+  var idlesAfter = this.options.idlesAfter * 60000;
 
-  // Write sessions can be used for reads and writes.
-  if (this.writePool && this.writePool.free) {
-    this.getWriteSession(callback);
-    return;
-  }
+  return this.available_.filter(function(session) {
+    return Date.now() - session.lastUsed >= idlesAfter;
+  });
+};
 
-  // Use a read session by calling begin transaction on it.
-  if (options.write && this.pool.free) {
-    this.getSession(function(err, session) {
-      if (err) {
-        callback(err);
-        return;
-      }
+/**
+ * Attempts to get a session of a specific type. If the type is unavailable it
+ * may try to use a different type.
+ *
+ * @private
+ *
+ * @param {string} type The desired session type.
+ * @returns {Promise<Session>}
+ */
+SessionPool.prototype.getSession_ = function(type) {
+  var self = this;
 
-      var transaction = self.createTransaction_(session);
+  if (!this.available_.length) {
+    if (this.options.fail) {
+      return Promise.reject(new Error('No resources available.'));
+    }
 
-      transaction.begin(function(err) {
-        if (err) {
-          callback(err);
-          return;
-        }
+    if (!this.isFull()) {
+      return this.race_(function() {
+        return self.createSession_(type);
+      });
+    }
 
-        callback(null, session, transaction);
+    return this.race_(function() {
+      return new Promise(function(resolve) {
+        self.once('available', resolve);
       });
+    }).then(function() {
+      return self.getSession_(type);
     });
-    return;
   }
 
-  // Use a regular read session.
-  if (this.pool.free) {
-    this.getSession(callback);
-    return;
+  var session;
+
+  for (session of this.available_) {
```

Comment:
why not maintain two lists, one for read sessions and one for read write sessions? Then you do not need to iterate through the list. You can just pop the head of the relevant list if non empty otherwise pop the other list.

Reply:
Good call, I've pushed a commit that does this.